source/JsMaterialX/test/XmlIo.spec.js

```diff
@@ -1,20 +1,243 @@
+            }
+            expect(doc.validate()).to.be.true;
+
+            // Make sure the document does actually contain something.
+            let valueElementCount = 0;
+            const treeIter = doc.traverseTree();
+            for(const elem of treeIter) {
+                if (elem instanceof mx.ValueElement) {
+                    valueElementCount++;
+                }
+            }
+            expect(valueElementCount).to.be.greaterThan(0);
+        };
+    }
+
+    before(async () => {
+        mx = await Module();
+    });
+
+    it('Read XML without includes from file', async () => {
+        // Read the standard library
+        const libs = await readStdLibrary(false);
+
+        // Read and validate the example documents.
+        readAndValidateExamples(examplesWithoutIncludes, libs,
+            async (document, file, sp) => {
+                await mx.readFromXmlFile(document, file, sp);
+            }, examplesPath);
+
+        // Read the same document twice, and verify that duplicate elements
+        // are skipped.
+        const doc = mx.createDocument();
+        const filename = 'CustomNode.mtlx';
+        await mx.readFromXmlFile(doc, filename, examplesPath);
+        await mx.readFromXmlFile(doc, filename, examplesPath);
+        expect(doc.validate()).to.be.true;
+    });
+
+    it('Read XML without includes from string', async () => {
+        // Read the standard library
+        const libs = await readStdLibrary(true);
+
+        // Read and validate each example document.
+        const examplesWithoutIncludesStrings = getMtlxStrings(examplesWithoutIncludes, examplesPath);
+        readAndValidateExamples(examplesWithoutIncludesStrings, libs,
+            async (document, file) => {
+                await mx.readFromXmlString(document, file);
+            });
+
+        // Read the same document twice, and verify that duplicate elements
+        // are skipped.
+        const doc = mx.createDocument();
+        const file = examplesWithoutIncludesStrings[examplesWithoutIncludes.indexOf('CustomNode.mtlx')];
+        await mx.readFromXmlString(doc, file);
+        await mx.readFromXmlString(doc, file);
+        expect(doc.validate()).to.be.true;
+    });
+
+    it('Read XML with includes', async () => {
+        const searchPath = libraryPath + ';' + examplesPath;
+
+        // Read the standard library
+        const libs = await readStdLibrary(false);
+
+        // Read and validate the example documents from files
+        readAndValidateExamples(examplesWithIncludes, libs,
+            async (document, file, sp) => {
+                await mx.readFromXmlFile(document, file, sp);
+            }, searchPath);
+
+        // Read and validate the example documents as strings
+        const examplesWithIncludesStrings = getMtlxStrings(examplesWithIncludes, searchPath);
+        readAndValidateExamples(examplesWithIncludesStrings, libs,
+            async (document, file, sp) => {
```

Comment:
Those these async functions work? The async part is inside the callback, you would need to add the await aswell in the readAndValidateExamples calls 🤔

Reply:
`readAndValidateExamples` has an `await` in front of the invoked callback.

Reply:
Ah, now I got what you mean. You're right. I guess it works because the code after the `readAndValidateExamples` calls doesn't depend on what happens inside of this function. I'll still add the `await`s, though.